rust/ql/lib/codeql/rust/dataflow/internal/DataFlowImpl.qll

```diff
@@ -586,16 +745,32 @@ module RustDataFlow implements InputSig<Location> {
    * are the value-preserving intra-callable flow steps.
    */
   predicate simpleLocalFlowStep(Node nodeFrom, Node nodeTo, string model) {
-    LocalFlow::localFlowStepCommon(nodeFrom, nodeTo) and
+    (
+      LocalFlow::localFlowStepCommon(nodeFrom, nodeTo)
+      or
+      exists(SsaImpl::DefinitionExt def, boolean isUseStep |
+        SsaFlow::localFlowStep(def, nodeFrom, nodeTo, isUseStep)
+      |
+        isUseStep = false
+        or
+        isUseStep = true and
+        not FlowSummaryImpl::Private::Steps::prohibitsUseUseFlow(nodeFrom, _)
+      )
```

Comment:
While it preserves semantics, it will introduce unnecessary tuple duplication. `a implies b` is syntactic sugar for `not a or b`, so the rewrite becomes
```
SsaFlow::localFlowStep(def, nodeFrom, nodeTo, isUseStep) and
not (isUseStep = true)
or
SsaFlow::localFlowStep(def, nodeFrom, nodeTo, isUseStep) and
not FlowSummaryImpl::Private::Steps::prohibitsUseUseFlow(nodeFrom, _)
```
Not how `isUseStep = true` is not part of the last disjunct.

Reply:
This is a bit hard for me to wrap my head around. If `isUseStep = true` is not the case then the first disjunct will always hold, so I would've thought that adding it to the last disjunct wouldn't make a difference. But I guess that's only the "preserves semantics" part.

Is it something along the lines of: For a disjunction `P or Q` all the tuples satisfying `P` will be created, then all tuples satisfying `Q` will be created, and only then will they be joined together (and de-duplicated). With my rewrite `Q` could be a larger stream of tuples overlapping with `P`, and hence the result will be the same but more de-duplication is going on?

I guess turning `def` into `_` is still applicable.

Reply:
> Is it something along the lines of: For a disjunction P or Q all the tuples satisfying P will be created, then all tuples satisfying Q will be created, and only then will they be joined together (and de-duplicated). With my rewrite Q could be a larger stream of tuples overlapping with P, and hence the result will be the same but more de-duplication is going on?

Exactly.

Reply:
I oversaw the `_` part; that is certainly valid, let's do it follow-up.